### PR TITLE
document Import From Derivation (IFD) policy

### PR DESCRIPTION
**Pull request by fricklerhandwerk**
###### Description of changes

document Nixpkgs policy on IFD and update existing link to the definition to point to this manual instead of the NixOS Wiki.

###### Things done

- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: doc/languages-frameworks/javascript.section.md [hunk 1]
```diff
@@ -330,21 +330,25 @@ mkYarnPackage rec {
   - The `echo 9` steps comes from this answer: <https://stackoverflow.com/a/49139496>
   - Exporting the headers in `npm_config_nodedir` comes from this issue: <https://github.com/nodejs/node-gyp/issues/1191#issuecomment-301243919>
 
-## Outside of nixpkgs {#javascript-outside-nixpkgs}
+## Outside of Nixpkgs {#javascript-outside-nixpkgs}
 
-There are some other options available that can't be used inside nixpkgs. Those other options are written in Nix. Importing them in nixpkgs will require moving the source code into nixpkgs. Using [Import From Derivation](https://nixos.wiki/wiki/Import_From_Derivation) is not allowed in Hydra at present. If you are packaging something outside nixpkgs, those can be considered
+There are some other tools available, which are written in the Nix language.
+These that can't be used inside Nixpkgs because they require [Import From Derivation](#ssec-import-from-derivation), which is not allowed in Nixpkgs.
+Using them in Nixpkgs requires checking the intermediate build results into version control.
```
Comment (figsoda) [inline doc/languages-frameworks/javascript.section.md:337]: I don't think it's possible to use them at all unless they are copy-pasted into nixpkgs
```suggestion
```

Comment (winterqt) [reply doc/languages-frameworks/javascript.section.md:337]: A more accurate/clear description instead of intermediate would be "files from the source," I think.

Comment (figsoda) [reply doc/languages-frameworks/javascript.section.md:337]: By `they` I meant to say npmlock2nix and nix-npm-buildpackage

Comment (winterqt) [reply doc/languages-frameworks/javascript.section.md:337]: Right, but that's not related to the section I highlighted, unless I'm misunderstanding.

I'm saying that we can keep it if we reword it to be more accurate/clear.

Comment (fricklerhandwerk) [reply doc/languages-frameworks/javascript.section.md:337]: @winterqt not sure what you mean there. How would you reword the sentence?

@figsoda why would one have to paste the tools themselves into Nixpkgs? They could be a build dependency like any other, and come from anywhere. The problem here is that they will produce files which will have to be read during evaluation, which constitutes IFD. To work around that, you'll have to paste *those files* into Nixpkgs and check them into version control.

Comment (figsoda) [reply doc/languages-frameworks/javascript.section.md:337]: IIUC both libraries are written in nix and you can't use them without ifd in nixokgs

Comment (fricklerhandwerk) [reply doc/languages-frameworks/javascript.section.md:337]: Yes, that is exactly what the blurb is trying to say. If the message doesn't come across I'm happy to accept suggestions.

Comment (winterqt) [reply doc/languages-frameworks/javascript.section.md:337]: I feel like we're all talking about different things here.

@figsoda is talking about how we can't use them from their VCS repos without IFD, which is true -- but nothing would stop us from copying them over to here, and developing them further here.

I'm talking about how both mentioned tools require the use of IFD anyways, because they depend on parsing files in the source repository to generate the required Nix expressions. I think my angle is the better one to specify in the docs (over the "intermediate build artifacts" wording), since it's not something we can work around without copying source files from all packages.

I hope that makes sense 😅

File: doc/contributing/coding-conventions.chapter.md [hunk 2]
```diff
@@ -670,3 +670,18 @@ stdenv.mkDerivation {
...
 ```
+
+### Import From Derivation {#ssec-import-from-derivation}
+
+Import From Derivation (IFD) is disallowed in Nixpkgs for performance reasons:
+[Hydra] evaluates the entire package set, and sequential builds during evalation would increase evaluation times to become impractical.
+
+[Hydra]: https://github.com/NixOS/hydra
+
+Import From Derivation can be worked around in some cases by checking light-weight intermediate build results, such as files containing dependency information, into version control and using them like regular sources.
```
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply doc/contributing/coding-conventions.chapter.md:681]: I think this obscures more than the original wording.

How about that:

```suggestion
Import From Derivation can be worked around in some cases by checking generated intermediate files into version control and reading those like regular sources instead.
```

[the remark this answers is not in the document]
Comment (roberth) [reply doc/contributing/coding-conventions.chapter.md:681]: > by checking

This reads like _checking_ instead of _checking into_.

```suggestion
Import From Derivation can be worked around in some cases by committing generated intermediate files and reading those like regular sources instead.
```

[the remark this answers is not in the document]
Comment (roberth) [reply doc/contributing/coding-conventions.chapter.md:681]:
```suggestion
Import From Derivation can be worked around in some cases by committing generated intermediate files and reading those instead.
```

To what degree are they like regular sources? "[committed] generated intermediate files" is descriptive enough, without provoking unproductive thoughts about the comparison.

File: doc/languages-frameworks/javascript.section.md (hunk 1 above)
Comment (fricklerhandwerk) [reply doc/languages-frameworks/javascript.section.md:337]: Ah okay, sorry. I was unaware that there is no explicit intermediate step that would allow avoiding IFD. Removed the sentence as suggested. I think it's irrelevant here that using them without IFD would require a fork. Well, of course, anything can be done if one rewrites the tool.

I hope this is correct now?

File: doc/contributing/coding-conventions.chapter.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply doc/contributing/coding-conventions.chapter.md:681]: Right. I was just stumbling over just saying "commit", because that's jargon and [appears to be non-grammatical](https://dictionary.cambridge.org/dictionary/english/commit) without a target.

```suggestion
Import From Derivation can be worked around in some cases by committing generated intermediate files to version control and reading those instead.
```

[the remark this answers is not in the document]
Comment (roberth) [reply doc/contributing/coding-conventions.chapter.md:681]: LGTM!

Explaining commit as jargon shows your commitment to accessibility. :heart:

**Review by figsoda — APPROVED** [review]

**Review by roberth — APPROVED** [review]
... but otherwise LGTM.

File: doc/languages-frameworks/javascript.section.md [hunk 3]
```diff
@@ -330,21 +330,24 @@ mkYarnPackage rec {
   - The `echo 9` steps comes from this answer: <https://stackoverflow.com/a/49139496>
   - Exporting the headers in `npm_config_nodedir` comes from this issue: <https://github.com/nodejs/node-gyp/issues/1191#issuecomment-301243919>
 
-## Outside of nixpkgs {#javascript-outside-nixpkgs}
+## Outside Nixpkgs {#javascript-outside-nixpkgs}
 
-There are some other options available that can't be used inside nixpkgs. Those other options are written in Nix. Importing them in nixpkgs will require moving the source code into nixpkgs. Using [Import From Derivation](https://nixos.wiki/wiki/Import_From_Derivation) is not allowed in Hydra at present. If you are packaging something outside nixpkgs, those can be considered
+There are some other tools available, which are written in the Nix language.
```
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply doc/languages-frameworks/javascript.section.md:335]: Yes, that's probably better. Will do that and merge what's left.

File: doc/contributing/coding-conventions.chapter.md [hunk 4]
```diff
@@ -670,3 +670,18 @@ stdenv.mkDerivation {
   ...
 }
 ```
+
+### Import From Derivation {#ssec-import-from-derivation}
+
+Import From Derivation (IFD) is disallowed in Nixpkgs for performance reasons:
+[Hydra] evaluates the entire package set, and sequential builds during evalation would increase evaluation times to become impractical.
```
Comment (dit7ya) [inline doc/contributing/coding-conventions.chapter.md:677]:
```suggestion
[Hydra] evaluates the entire package set, and sequential builds during evaluation would increase evaluation times to become impractical.
```

Comment (fricklerhandwerk) [reply doc/contributing/coding-conventions.chapter.md]: Thanks, good catch.

**Review by roberth — APPROVED** [review]
